File: 035-Search-Insert-Position/solution02.py
#!/usr/bin/python
# ==============================================================================
# Author:   Tao Li (taoli@ucsd.edu)
# Date:     May 7, 2015
# Question: 035-Search-Insert-Position
# Link:     https://leetcode.com/problems/search-insert-position/
# ==============================================================================
# Given a sorted array and a target value, return the index if the target is 
# found. If not, return the index where it would be if it were inserted in order.
# 
# You may assume no duplicates in the array.
# 
# Here are few examples.
# [1,3,5,6], 5 -> 2
# [1,3,5,6], 2 -> 1
# [1,3,5,6], 7 -> 4
# [1,3,5,6], 0 -> 0
# ==============================================================================
# Method: Binary Search
# Time Complexity: O(logN)
# Space Complexity: O(N)
# ==============================================================================

class Solution:
    # @param {integer[]} nums
    # @param {integer} target
    # @return {integer}
    def searchInsert(self, nums, target):
        low = 0
        high = len(nums)-1

        while low <= high:
            mid = low + (high-low)/2
            if target > nums[mid]:
                low = mid + 1
            elif target < nums[mid]:
                high = mid - 1
            else:
                return mid
        return low

if __name__ == '__main__':
    print Solution().searchInsert([1,3,5,6], 5)

File: 033-Search-in-Rotated-Sorted-Array/CPP/Solution02.cpp
/*
==============================================================================
Author:   Tao Li (taoli@ucsd.edu)
Date:     Jul 9, 2015
Question: 033-Search-in-Rotated-Sorted-Array
Link:     https://leetcode.com/problems/search-in-rotated-sorted-array/
==============================================================================
Suppose a sorted array is rotated at some pivot unknown to you beforehand.

(i.e., 0 1 2 4 5 6 7 might become 4 5 6 7 0 1 2).

You are given a target value to search. If found in the array return its 
index, otherwise return -1.

You may assume no duplicate exists in the array.
==============================================================================
Method: brute force
Time Complexity: O(n)
Space Complexity: O(1)
==============================================================================
*/

#include <vector>
#include <iostream>
using namespace std;

class Solution {
public:
    int search(vector<int>& nums, int target) {
        int i = 0;
        int length = nums.size();
        while (i < length) {
            if (nums[i] == target) break;
            i ++;
        }
        return i < length ? i : -1;
    }
};

int main(int argc, char const *argv[])
{
    /* code */
    int a[5] = {1,2,3,4,5};
    vector<int> v (a, a+5);
    Solution s;
    cout << s.search(v, 10) << endl;
    return 0;
}